orttraining/orttraining/training_ops/cuda/nn/dropout.cc

```diff
@@ -102,5 +102,112 @@ Status DropoutGrad<T1, T2>::ComputeInternal(OpKernelContext* context) const {
+  ONNX_OPERATOR_TYPED_KERNEL_EX(                                     \
+      BiasDropout,                                                   \
+      kMSDomain,                                                     \
+      1,                                                             \
+      T1##_##T2,                                                     \
+      kCudaExecutionProvider,                                        \
+      KernelDefBuilder()                                             \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<T1>())    \
+          .TypeConstraint("T1", DataTypeImpl::GetTensorType<T2>())   \
+          .TypeConstraint("T2", DataTypeImpl::GetTensorType<bool>()) \
+          .InputMemoryType<OrtMemTypeCPUInput>(3)                    \
+          .InputMemoryType<OrtMemTypeCPUInput>(4),                   \
+      BiasDropout<T1, T2>);
+
+REGISTER_BIAS_DROPOUT_KERNEL_TYPED(MLFloat16, MLFloat16)
+REGISTER_BIAS_DROPOUT_KERNEL_TYPED(MLFloat16, float)
+REGISTER_BIAS_DROPOUT_KERNEL_TYPED(float, MLFloat16)
+REGISTER_BIAS_DROPOUT_KERNEL_TYPED(float, float)
+
+template <typename T1, typename T2>
+Status BiasDropout<T1, T2>::ComputeInternal(OpKernelContext* context) const {
+  typedef typename ToCudaType<T1>::MappedType CudaT;
+ 
+  //Get X_data
+  const Tensor* X = context->Input<Tensor>(0);
+  if (X == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "X Input is not available.");
+  const TensorShape& x_shape = X->Shape();
+  auto X_data = reinterpret_cast<const CudaT*>(X->template Data<T1>());
+  const int64_t N = x_shape.Size();
+
+ //Get bias_data
+  const Tensor* bias = context->Input<Tensor>(1);
+  if (bias == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "Bias input of BiasDropout is not available.");
+  const TensorShape& bias_shape = bias->Shape();
+  if (bias_shape.NumDimensions() != 1) {
+    return Status(common::ONNXRUNTIME, common::FAIL, "Bias input is not a 1D tensor.");
+  }
+  const int64_t dim = bias_shape[0];
+  if (dim != x_shape.GetDims().back()){
+    return Status(common::ONNXRUNTIME, common::FAIL, "Bias' dimension doesn't match input's last dimension.");
+  }
+  auto bias_data = reinterpret_cast<const CudaT*>(bias->template Data<T1>());
+
+  //Get residual_data
+  const Tensor* residual = context->Input<Tensor>(2);
+  const CudaT* residual_data = nullptr;
+  if (residual != nullptr) {
+    const TensorShape& residual_shape = residual->Shape();
+    if (residual_shape != x_shape) {
+      return Status(common::ONNXRUNTIME, common::FAIL, "Residual input shape does not match X input shape.");
+    }
+    residual_data = reinterpret_cast<const CudaT*>(residual->template Data<T1>());
+  }
+
+  //Get Y_data
+  auto Y = context->Output(0, x_shape);
+  auto Y_data = reinterpret_cast<CudaT*>(Y->template MutableData<T1>());
+
+  //Get mask_data
+  auto mask = context->Output(1, x_shape);
+  ORT_ENFORCE(!mask || mask->Shape().Size() == N);
```

Comment:
>ORT_ENFORCE [](start = 2, length = 11)

consider adding messages to ORT_ENFORCE's

Reply:
This is actually a useless check... removed.

---
In reply to: [441072696](https://github.com/microsoft/onnxruntime/pull/4167#discussion_r441072696) [](ancestors = 441072696)